src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Cohost/CohostEndpointTest.cs

```diff
@@ -0,0 +1,186 @@
+            .AddExcludedPartTypes(typeof(IWorkspaceProvider))
+            .AddParts(typeof(TestWorkspaceProvider))
+            .AddExcludedPartTypes(typeof(ILspEditorFeatureDetector))
+            .AddParts(typeof(TestLspEditorFeatureDetector));
+
+        using var exportProvider = testComposition.ExportProviderFactory.CreateExportProvider();
+
+        var providers = exportProvider.GetExportedValues<IDynamicRegistrationProvider>().ToList();
+
+        // First we verify that the MEF composition above is correct, otherwise this test will be invalid
+        var actualProviders = typeof(CohostLinkedEditingRangeEndpoint).Assembly.GetTypes().Where(t => !t.IsInterface && typeof(IDynamicRegistrationProvider).IsAssignableFrom(t)).ToList();
+        Assert.Equal(actualProviders.OrderBy(a => a.Name).Select(r => r.Name).ToArray(), providers.OrderBy(e => e.GetType().Name).Select(r => r.GetType().Name).ToArray());
+
+        var clientCapabilities = new VSInternalClientCapabilities()
+        {
+            SupportsVisualStudioExtensions = true,
+            TextDocument = new TextDocumentClientCapabilities()
+            {
+                CodeAction = new() { DynamicRegistration = true },
+                CodeLens = new() { DynamicRegistration = true },
+                Completion = new() { DynamicRegistration = true },
+                Definition = new() { DynamicRegistration = true },
+                Diagnostic = new() { DynamicRegistration = true },
+                DocumentHighlight = new() { DynamicRegistration = true },
+                DocumentLink = new() { DynamicRegistration = true },
+                DocumentSymbol = new() { DynamicRegistration = true },
+                FoldingRange = new() { DynamicRegistration = true },
+                Formatting = new() { DynamicRegistration = true },
+                Hover = new() { DynamicRegistration = true },
+                Implementation = new() { DynamicRegistration = true },
+                InlayHint = new() { DynamicRegistration = true },
+                LinkedEditingRange = new() { DynamicRegistration = true },
+                OnTypeFormatting = new() { DynamicRegistration = true },
+                RangeFormatting = new() { DynamicRegistration = true },
+                References = new() { DynamicRegistration = true },
+                Rename = new() { DynamicRegistration = true },
+                SemanticTokens = new() { DynamicRegistration = true },
+                SignatureHelp = new() { DynamicRegistration = true },
+                Synchronization = new() { DynamicRegistration = true },
+                TypeDefinition = new() { DynamicRegistration = true }
+            },
+        };
+
+        foreach (var endpoint in providers)
+        {
+            if (endpoint is CohostSemanticTokensRangeEndpoint)
+            {
+                // We can't currently test this, as the GetRegistrations method calls requestContext.GetRequiredService
```

Comment:
What about a Mock?

Reply:
Can't mock structs, and RazorCohostRequestContext is a struct, which takes a roslyn RequestContext as a constructor parameter, which is also a struct :)

Reply:
ah I missed it was a struct!